cmd/node-agent/main.go

```diff
@@ -0,0 +1,277 @@
+		},
+	)
+	var configMapController *framework.Controller
+	configMapStore, configMapController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(lo api.ListOptions) (runtime.Object, error) {
+				return client.Core().ConfigMaps(api.NamespaceSystem).List(lo)
+			},
+			WatchFunc: func(lo api.ListOptions) (watch.Interface, error) {
+				return client.Core().ConfigMaps(api.NamespaceSystem).Watch(lo)
+			},
+		},
+		&v1.ConfigMap{},
+		30*time.Second,
+		framework.ResourceEventHandlerFuncs{},
+	)
+
+	go configMapController.Run(wait.NeverStop)
+	nodeController.Run(wait.NeverStop)
+}
+
+func newAPIClient() clientset.Interface {
+	config, err := restclient.InClusterConfig()
+	if err != nil {
+		glog.Fatal(err)
+	}
+	return clientset.NewForConfigOrDie(config)
+}
+
+func nodeUpdateCallback(oldObj, newObj interface{}) {
+	glog.Infof("begin node update callback")
+
+	newNode, ok := newObj.(*v1.Node)
+	if !ok {
+		glog.Errorf("received unexpected type: %T (expected *v1.Node)", newObj)
+		return
+	}
+	if newNode.Annotations == nil {
+		newNode.Annotations = make(map[string]string)
+	}
+	rawconf, ok := newNode.Annotations[desiredConfigAnnotation]
+	if !ok {
+		glog.Info("no configuration annotation found, ignoring")
+		return
+	}
+	var desiredConf map[string]string
+	err := json.Unmarshal([]byte(rawconf), &desiredConf)
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+
+	onDiskConfig, err := parseKubeletEnvFile(kubeletEnvPath)
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+	cm, err := getConfigMap(desiredConf[kubeletConfigKey])
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+	if configHasChanged(onDiskConfig, desiredConf, cm) {
+		if err := validateConfig(desiredConf); err != nil {
+			glog.Error(err)
+			return
+		}
+		if err := updateKubeletEnvFile(desiredConf, onDiskConfig, cm); err != nil {
+			glog.Error(err)
+			return
+		}
+		if err := restartKubeletService(); err != nil {
+			glog.Error(err)
+			return
+		}
+	}
+	// Always update annotation at end of sync loop.
+	if err := updateNodeAnnotation(newNode, desiredConf); err != nil {
+		glog.Error(err)
+	}
+}
+
+func configHasChanged(onDiskConfig, desiredConfig map[string]string, cm *v1.ConfigMap) bool {
+	if onDiskConfig[kubeletVersionKey] != desiredConfig[kubeletVersionKey] {
+		return true
+	}
+	if onDiskConfig[kubeletFlagsKey] != cm.Data[configMapFlagsKey] {
+		return true
+	}
+	return false
+}
+
+func validateConfig(desiredConfig map[string]string) error {
+	if _, ok := desiredConfig[kubeletVersionKey]; !ok {
+		return fmt.Errorf("configuration annotation does not contain required key: %s", kubeletVersionKey)
+	}
+	if _, ok := desiredConfig[kubeletConfigKey]; !ok {
+		return fmt.Errorf("configuration annotation does not contain required key: %s", kubeletConfigKey)
+	}
+	return nil
+}
+
+func updateKubeletEnvFile(conf, onDiskConfig map[string]string, cm *v1.ConfigMap) error {
+	var buf bytes.Buffer
+	onDiskConfig[kubeletVersionKey] = conf[kubeletVersionKey]
```

Comment:
nit: updating the onDiskConfig map as a side effect of this function feels a bit odd. Or if we do, it should be updated on success or return a new copy.

If the goal is to have the env file match our desired config exactly, we can probably just drop this merge and not pass in onDiskConfig at all (just construct the data you want on disk from conf/cm)